<!DOCTYPE html>
<html>
  <head>
    <title>Scared no more:</title>
    <meta charset="utf-8">
    <meta name="author" content="Anna Quaglieri" />
    <link href="MiltonMyFriend_files/font-awesome/css/fontawesome-all.min.css" rel="stylesheet" />
    <link rel="stylesheet" href="default.css" type="text/css" />
    <link rel="stylesheet" href="MUG-fonts.css" type="text/css" />
  </head>
  <body>
    <textarea id="source">
class: center, middle, inverse, title-slide

# Scared no more:
## A beginner guide to make Milton <img src="img/Milton.png" /> your friendly collaborator
### Anna Quaglieri
### 2018/06/22

---





## Two most common types of workflows in üë®‚Äçüíª  Bioinformatics üë©‚Äçüíª:

&lt;br&gt;

* **Interactive Sessions**: We need to run code and see interactively what‚Äôs happening (plots, model results etc...). For example, by using the terminal or <i class="fab  fa-r-project "></i> Studio.
 
* **Serial/Pipeline jobs** üîÅ: We need to perform the same series of steps (alignment,remove adapters, sort,mark duplicated, etc...) for tens of samples and we care only about the final result.

---

## Today's talk ![](img/Milton.png)

&lt;br&gt;
&lt;br&gt;

* A few comments about **Interactive Sessions**
 
* Why did I start using **Milton**? 
 
* How I use **Milton**


---


## Interactive Sessions with <i class="fab  fa-r-project "></i>

### From your local <i class="fab  fa-r-project "></i> Studio

üëç I can update <i class="fab  fa-r-project "></i> üì¶ and version when I need to &lt;br&gt;
&lt;br&gt;
üëç I can access files üóÑ on the server and on my local üíª 
&lt;br&gt;


```r
# Beginning of every script
# TRUE if running from local, FALSE if running from the server
local &lt;- TRUE
mydir &lt;- ifelse(local,"/Volumes/quaglieri.a/PHD_project/GEO_Leucegene_data",
                "/home/users/allstaff/quaglieri.a/PHD_project/GEO_Leucegene_data")
```

üëé Slow üê¢ or even does not work with very large files

---

&lt;br&gt;
&lt;br&gt;

### <i class="fab  fa-r-project "></i> Studio on Unix or Milton


* http://unix401/auth-sign-in / http://unix309/auth-sign-in /  http://unix308/auth-sign-in
 
* I have never ran it using Milton but it is possible!
 
üëç Faster ‚ö° than running from local <i class="fab  fa-r-project "></i> Studio! &lt;br&gt;
&lt;br&gt;
üëç Cannot access directly files üóÑ on my local üíª &lt;br&gt;



---

&lt;br&gt;
&lt;br&gt;

### `ssh unix500`

* Best of both worlds ü•Ç: hybrid between `torquelord` and the well known `unix` computers
 
* You can run analyses directly from it like a normal unix terminal, you can open screen sessions and you can also submit jobs on torquelord with `qsub`

* [Watch video](https://drive.google.com/file/d/1v3OmRfSk6NLZiIBB0ce74obP6aZsNxkm/view?usp=sharing)!


---

## Why did I start using Milton


* I had to run the same pipeline over and over again across many samples
* I got tired of opening tens of screen sessions on the unix machines to run jobs in "parallel"


---

### RNA-Seq pipeline: FASTQ files -&gt; Annotated variants

Initial run for 46 RNA-Seq samples downloaded from GEO

1. FASTQC on raw Fastq files 
2. Alignment: `STAR` pass1 + `STAR` pass2
3. Mark Duplicates + add Read Groups with `Picard Tools`
5. Some QC: `ValidateSamFile`
4. Gene counts with `featureCounts`
5. Fusions with `STAR-Fusion`
6. GATK pre-processing (very slow!)
7. Call variants with four callers
8. Annotate variants with `VEP`


---

Pipeline to be run: 


```r
(46 * 3) + (4 * (46 * 5))
```

```
#&gt; [1] 1058
```

Times!

![](img/RNADownsRuns.png)


---

## How to use Milton: Example aligning reads with `Subread`

* We have to align 46 fastq files with `Subread::align()` in R
* Let's suppose we have created the `Subread` index since that needs to be created once for all!


```r
local &lt;- TRUE
mydir &lt;- ifelse(local,"/Volumes/quaglieri.a/PHD_project/GEO_Leucegene_data",
                "/home/users/allstaff/quaglieri.a/PHD_project/GEO_Leucegene_data")

library(Rsubread)

# Run Subread in the directory where the index is stored
setwd(file.path(mydir,"genomes/Subread_index"))
list.files()
```

```
#&gt;  [1] "GRCh38_index.00.b.array" "GRCh38_index.00.b.tab"  
#&gt;  [3] "GRCh38_index.01.b.array" "GRCh38_index.01.b.tab"  
#&gt;  [5] "GRCh38_index.02.b.array" "GRCh38_index.02.b.tab"  
#&gt;  [7] "GRCh38_index.files"      "GRCh38_index.log"       
#&gt;  [9] "GRCh38_index.reads"      "hg19_index.00.b.array"  
#&gt; [11] "hg19_index.00.b.tab"     "hg19_index.01.b.array"  
#&gt; [13] "hg19_index.01.b.tab"     "hg19_index.02.b.array"  
#&gt; [15] "hg19_index.02.b.tab"     "hg19_index.files"       
#&gt; [17] "hg19_index.log"          "hg19_index.reads"
```

---

### Run `Subread` on a terminal screen session

* Start terminal screen session


```bash
ssh unix308
screen
module load R
R
```

* Run `Subread`


```r
local &lt;- FALSE
mydir &lt;- ifelse(local,"/Volumes/quaglieri.a/PHD_project/GEO_Leucegene_data",
                "/home/users/allstaff/quaglieri.a/PHD_project/GEO_Leucegene_data")

library(Rsubread)

setwd(file.path(mydir,"genomes/Subread_index"))

# Get FASTQ files name
fastqdir &lt;- "/wehisan/general/user_managed/grpu_majewski_3/quaglieri.a/GEO_Leucegene_data/sra_cbf_aml/fastq_20M_100"
reads1 &lt;- list.files(path = fastqdir,pattern = "R1.fastq.gz$")
reads2 &lt;- gsub("R1","R2",reads1)

# Align
align(index="hg19_index",readfile1=reads1,readfile2=reads2,phredOffset=64,nthreads = 10)
```


---

### `Subread` threading

* `Subread` will align the fastq files one after the other one (like a `for` loop)

* It applies multithreading when aligning the reads within every fastq file

* Milton will help you to:

  * **Align every fasq files at the same time**
  
  * Use `Subread` multithreading for every fastq file
  

---

## Create
    </textarea>
<script src="https://remarkjs.com/downloads/remark-latest.min.js"></script>
<script>var slideshow = remark.create({
"highlightStyle": "github",
"highlightLines": true,
"countIncrementalSlides": false
});
if (window.HTMLWidgets) slideshow.on('afterShowSlide', function (slide) {
  window.dispatchEvent(new Event('resize'));
});
(function() {
  var d = document, s = d.createElement("style"), r = d.querySelector(".remark-slide-scaler");
  if (!r) return;
  s.type = "text/css"; s.innerHTML = "@page {size: " + r.style.width + " " + r.style.height +"; }";
  d.head.appendChild(s);
})();</script>

<script type="text/x-mathjax-config">
MathJax.Hub.Config({
  tex2jax: {
    skipTags: ['script', 'noscript', 'style', 'textarea', 'pre']
  }
});
</script>
<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
(function () {
  var script = document.createElement('script');
  script.type = 'text/javascript';
  script.src  = 'https://cdn.bootcss.com/mathjax/2.7.1/MathJax.js?config=TeX-MML-AM_CHTML';
  if (location.protocol !== 'file:' && /^https?:/.test(script.src))
    script.src  = script.src.replace(/^https?:/, '');
  document.getElementsByTagName('head')[0].appendChild(script);
})();
</script>
  </body>
</html>
