<!DOCTYPE html>
<html>
  <head>
    <title>Scared no more:</title>
    <meta charset="utf-8">
    <meta name="author" content="Anna Quaglieri" />
    <link href="Slides_MiltonMyFriend_files/font-awesome/css/fontawesome-all.min.css" rel="stylesheet" />
    <link rel="stylesheet" href="default.css" type="text/css" />
    <link rel="stylesheet" href="MUG-fonts.css" type="text/css" />
  </head>
  <body>
    <textarea id="source">
class: center, middle, inverse, title-slide

# Scared no more: <img src="img/Milton.png" />
## A beginner guide to make Milton your friendly collaborator
### Anna Quaglieri
### 2018/06/22

---





## Two most common types of workflows in üë®‚Äçüíª  Bioinformatics üë©‚Äçüíª:

&lt;br&gt;

* **Interactive Sessions**: We need to run code and see interactively what‚Äôs happening (plots, model results etc...). For example, by using the terminal or <i class="fab  fa-r-project "></i> Studio.
 
* **Serial/Pipeline jobs** üîÅ: We need to perform the same series of steps (alignment,remove adapters, sort,mark duplicated, etc...) for tens of samples and we care only about the final result.

---

## Today's talk 

&lt;br&gt;
&lt;br&gt;

* A few comments about **Interactive Sessions**
 
* Why did I start using **Milton**? 
 
* How I use **Milton**


---


## Interactive Sessions with <i class="fab  fa-r-project "></i>

### From your local <i class="fab  fa-r-project "></i> Studio

üëç I can update <i class="fab  fa-r-project "></i> üì¶ and version when I need to &lt;br&gt;
&lt;br&gt;
üëç I can access files üóÑ on the server and on my local üíª 
&lt;br&gt;


```r
# Beginning of every script
# TRUE if running from local, FALSE if running from the server
local &lt;- TRUE
mydir &lt;- ifelse(local,"/Volumes/quaglieri.a/PHD_project",
                "/home/users/allstaff/quaglieri.a/PHD_project")
```

üëé Slow üê¢ or even does not work with very large files

---

&lt;br&gt;
&lt;br&gt;

### <i class="fab  fa-r-project "></i> Studio on Unix or Milton


* http://unix401/auth-sign-in / http://unix309/auth-sign-in /  http://unix308/auth-sign-in
 
* I have never ran it using Milton but it is possible!
 
üëç Faster ‚ö° than running from local <i class="fab  fa-r-project "></i> Studio! &lt;br&gt;
&lt;br&gt;
üëç Cannot access directly files üóÑ on my local üíª &lt;br&gt;



---

&lt;br&gt;
&lt;br&gt;

### `ssh unix500`

* Best of both worlds ü•Ç: hybrid between `torquelord` and the well known `unix` computers
 
* You can run analyses directly from it like a normal unix terminal, you can open screen sessions and you can also submit jobs on torquelord with `qsub`

---

## Message from <i class="fab  fa-r-project "></i> users to IT

&lt;br&gt;
&lt;br&gt;

* It would be great that when <i class="fab  fa-r-project "></i> is updated on the servers, it is also updated on <i class="fab  fa-r-project "></i> Studio on the servers. 

* The latest <i class="fab  fa-r-project "></i> version should be the default everywhere and if someone needs a previous version will go through the terminal

---

## Why did I start using Milton

&lt;br&gt;
&lt;br&gt;
&lt;br&gt;

* I had to run the same pipeline over and over again across many samples

* I got tired of opening tens of screen sessions on the unix machines to run jobs in "parallel"

---

### RNA-Seq pipeline: FASTQ files -&gt; Annotated variants

Initial run for 46 RNA-Seq samples downloaded from GEO

1. FASTQC on raw Fastq files 
2. Alignment: `STAR` pass1 + `STAR` pass2
3. Mark Duplicates + add Read Groups with `Picard Tools`
5. Some QC: `ValidateSamFile`
4. Gene counts with `featureCounts`
5. Fusions with `STAR-Fusion`
6. GATK pre-processing (very slow!)
7. Call variants with four callers
8. Annotate variants with `VEP`


---

Pipeline to be run: 


```r
(46 * 3) + (4 * (46 * 5))
```

```
#&gt; [1] 1058
```

Times!

![](img/RNADownsRuns.png)

---

## How I use Milton

&lt;br&gt;
&lt;br&gt;

1. Final structure of `job.sh` to submit with `qsub`

2. How to create an R script for a job (many different ways really!)

3. How to submit the R script to the queue system

4. How to monitor how your job is doing and where it is at

5. Check statistics after job is run to improve your submission

---

## Final structure of job to submit

Really, you can create it in the way you prefer but the final structure should look like the one below!

![](img/script_example.png)

---

&lt;br&gt;
&lt;br&gt;


```bash
ssh unix500

qsub ScriptToSubmit.sh
```

![](img/example_job_running.png)

* `NDS` = `nodes=1`

* `TSK` = `ppn=4`

* `Req'd memory` = `mem=4gb`

* `Req'd time` = `walltime=hh:mm:ss` see [here](http://catalyst.wehi.edu.au/research/computationalbiology/Pages/walltimes.aspx) fore info about walltime 

Find more information on the Milton page on [Catalyst](http://catalyst.wehi.edu.au/research/computationalbiology/Pages/Quick-Start-Guide-to-using-the-HPC.aspx)


---

&lt;br&gt;
&lt;br&gt;
&lt;br&gt;

Check example script `AlignWithSubread.Rmd` on  

<i class="fab  fa-github "></i> https://github.com/annaquaglieri16/MiltonMyFriend


```bash
git clone https://github.com/annaquaglieri16/MiltonMyFriend
```

---

## Summary of key commands that you want to keep in mind

* **Submit a job**: `qsub` (see slide above for `#PBS` specifications)

* **Check running jobs**: `qstat -n1 -u $USER`

* `ssh trqn-###` : ssh into a computer (node) where you can use `top` to monitor your job

* `cd /stornext/HPCScratch/.torque/spool/` to check `stdout` and `stderr` of running jobs. Use `Jobid` to find a specific job

* **Delete a job already submitted**: `qdel jobid`  (JOBID.torquelord3.hps0080)

* `module load its-tools` and `torquestats -u $USER -p` to monitor usage of past jobs and find all your saved `.csv` stats here `/stornext/System/data/torque_data/`


---

# Thanks to Helpdesk!

![](img/helpdesk.png)
    </textarea>
<script src="https://remarkjs.com/downloads/remark-latest.min.js"></script>
<script>var slideshow = remark.create({
"highlightStyle": "github",
"highlightLines": true,
"toc": true,
"countIncrementalSlides": false
});
if (window.HTMLWidgets) slideshow.on('afterShowSlide', function (slide) {
  window.dispatchEvent(new Event('resize'));
});
(function() {
  var d = document, s = d.createElement("style"), r = d.querySelector(".remark-slide-scaler");
  if (!r) return;
  s.type = "text/css"; s.innerHTML = "@page {size: " + r.style.width + " " + r.style.height +"; }";
  d.head.appendChild(s);
})();</script>

<script type="text/x-mathjax-config">
MathJax.Hub.Config({
  tex2jax: {
    skipTags: ['script', 'noscript', 'style', 'textarea', 'pre']
  }
});
</script>
<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
(function () {
  var script = document.createElement('script');
  script.type = 'text/javascript';
  script.src  = 'https://cdn.bootcss.com/mathjax/2.7.1/MathJax.js?config=TeX-MML-AM_CHTML';
  if (location.protocol !== 'file:' && /^https?:/.test(script.src))
    script.src  = script.src.replace(/^https?:/, '');
  document.getElementsByTagName('head')[0].appendChild(script);
})();
</script>
  </body>
</html>
